[Norman] iP

normkoh opened the pull request:
# DukePro
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is:
1. Download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😳 

AND it is **FREE**!

Features: 

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)  

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

normkoh committed:
Implemented greet, echo, exit.

normkoh committed:
Implemented Add, List.

normkoh committed:
Implemented Mark as Done. Added Task class.

normkoh committed:
Implemented ToDos, Deadlines and Events

normkoh committed:
Implemented automated  TextUiTesting.

normkoh committed:
Implemented basic error handling.

normkoh committed:
Implemented delete.

normkoh committed:
Add save function

normkoh committed:
Store deadline and event dates as LocalDate

normkoh committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java

normkoh committed:
Use more OOP

normkoh committed:
Move all classes to duke package

normkoh committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

normkoh committed:
Add JUnit tests

normkoh committed:
Fix some bugs

normkoh committed:
Write javadocs for some classes

normkoh committed:
Add JavaDocs for some classes

normkoh committed:
Add find function

normkoh committed:
Revert "Write javadocs for some classes"

This reverts commit 01f710708637700b370ccf037f3f8cd7a813fb7a.

normkoh committed:
Change code to fit coding standards

normkoh committed:
Merge branch 'branch-A-JavaDoc' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/duke/Ui.java

normkoh committed:
Merge branch 'branch-A-CodingStandard'

normkoh committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Ui.java

normkoh committed:
Add GUI

normkoh committed:
Add checkstyle

normkoh committed:
Add additional JavaDoc

normkoh committed:
Add JavaDoc

Thing1Thing2 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + by.format(DateTimeFormatter.ofPattern("MMM-d-yyyy")) + ")";
```
Should this be shorter?

Thing1Thing2 commented on src/main/java/duke/MarkCommand.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        if (isMark) {
+            tasks.markTask(taskNumber);
+            storage.save(tasks);
+            return ui.showMarkTask();
+        } else {
+            tasks.unmarkTask(taskNumber);
+            storage.save(tasks);
+            return ui.showUnmarkTask();
+        }
+    }
```
MarkCommand is simplifying the marking process as a toggle.
This is a nice way of avoiding two classes for Mark and Unmark! 👍

Thing1Thing2 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,79 @@
+     * @return Command based on user input.
+     * @throws DukeException the exception encountered while parsing through user input.
+     */
+    public Command parseCommand(String commandString) throws DukeException {
```
I like the code quality here, it is easy to follow how commandString is parsed. 😎

Thing1Thing2 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,88 @@
+                    String nextLine = sc.nextLine();
+                    String[] taskInfo = nextLine.split(" \\| ");
+                    Task curr = null;
+                    switch (taskInfo[0]) {
```
Perhaps an enum to make it easier to understand index of information stored in taskInfo might be needed.

Thing1Thing2 commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,32 @@
-Hello from
```
Think it is okay to delete text-ui-test now 🤔

Thing1Thing2 reviewed (commented):
Hi, it was nice reading your code for Duke.
It is quite succinct and easy to follow.
The cat-spin you added was fun :smile_cat:

I would suggest creating separate packages for the classes, since they are all inside the duke package currently :thought_balloon:

PS: I used your code to help resolve issues with adding GUI in my own. Thank you! :partying_face:

lyuting47 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,137 @@
+     * @param tl the taskList to be shown to the user.
+     * @return String containing display message to user.
+     */
+    public String showList(TaskList tl) {
+        StringBuilder sb = new StringBuilder();
+        List<Task> taskList = tl.getTasks();
+        int index = 1;
+        sb.append("Here are the tasks in your list nya:\n");
+        for (Task t : taskList) {
+            sb.append(index + "." + t + "\n");
+            index++;
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Displays the current taskList to the user.
+     *
+     * @param tl the taskList to be shown to the user.
+     * @param filter the filter to find tasks by.
+     * @return String containing display message to user.
+     */
+    public String showList(TaskList tl, String filter) {
+        StringBuilder sb = new StringBuilder();
+        List<Task> taskList = tl.getTasks();
+        int index = 1;
+        sb.append("Here are the matching tasks in your list nya:\n");
+        for (Task t : taskList) {
+            if (t.description.contains(filter)) {
+                sb.append(index + "." + t + "\n");
+                index++;
+            }
+        }
+        return sb.toString();
```
Is it better to move the logic of getting the relevant tasks to TaskList class instead? So as to follow the "tell, don't ask" principle.

lyuting47 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Represents the commands that Duke can accept to do various tasks
+ */
+public abstract class Command {
```
Abstracting out the Command class reduces a lot of if else's in the Parser class and makes it really clean. Nice!

lyuting47 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * Represents a Task that can be stored by Duke
+ */
+public abstract class Task {
```
Is there a need to make Task an abstract class? It has many default implementations of various methods, and perhaps in future we want to allow the user to add a generic task to Duke.

lyuting47 reviewed (commented):
Looks good! The code is easy to read and follows the given coding standards pretty closely. Only suggested minor changes to follow OO principles.

normkoh committed:
Add assertions

Code has no assertions implemented.

Assertions help to check certain parts of the code for certain
assumed behaviour.

Let's,
* add an assertion to check if program exits only when exitCommand
  is executed

normkoh committed:
Improve code quality

normkoh committed:
Merge pull request #1 from normkoh/branch-A-Assertions

Add assertions

normkoh committed:
Merge branch 'master' into branch-A-CodeQuality

normkoh committed:
Merge pull request #2 from normkoh/branch-A-CodeQuality

Improve code quality

normkoh committed:
Add detection of duplicate entries

No detection of duplicate tasks added to list.

Users may mistakenly add duplicate entries and it would help to
inform them of this to prevent task list from cluttering too much.

Let's,
* add ability for Duke to detect duplicate entries entered by user
* notify the user when a duplicate task is being entered

normkoh committed:
Fix bugs regarding detection of duplicates

normkoh committed:
Add user guide

normkoh committed:
Fix javadoc issues
